kep/115-Subgroup-support/README.md

```diff
@@ -0,0 +1,201 @@
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+To avoid the changes from affecting initial functionality and to not set 
+unnecessary labels, there will be no default behavior if `SubGroupSize` is not 
+set
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+### LeaderWorkerSet API
+We extend the LeaderWorkerSet API to introduce a new field: subGroupSize to opt in and set the number of pods that each subgroup will contain. Current behavior is kept if not set. 
+
+```
+type LeaderWorkerSetSpec struct {
+	 // Number of pods per subgroup
+	SubGroupSize *int32 'json:"subGroupSize,omitempty'
+} 
+```
+
+### Exclusive Topology Support
+LeaderWorkerSet can guarantee that the leader and the workers are placed in the same topology if the leaderworkerset.gke.io/exclusive-topology annotation is set. Similarly, we will support that the pods within the same subgroup will be placed in the same topology with a new annotation leaderworkerset.gke.io/exclusive-topology-subgroup, so that the new changes can support up to two levels of pod affinity. 
+
+### Implementation
+- Two new annotations will be added
+  - `leaderworkerset.gke.io/exclusive-topology-subgroup `
+  - `leaderworkerset.gke.io/subgroup-size` 
+- Three new labels will be added,
+  - `leaderworkerset.gke.io/subgroup-index = worker-index/subGroupSize`
+    - Tracks which subgroup the pod is part of 
+  - `leaderworkerset.gke.io/subgroup-worker-index = worker-index%subGroupSize`
+    - index/identity of a pod inside the pod's subgroup
+  - `leaderworkerset.gke.io/subgroup-key` 
+    - Pods that are part of the same subgroup will have an annotation that is a unique hash value will be generated from the name of the leader, and the subgroup-index
+
+To support exclusive placement at the subgroup level, the pod webhook will inject the new labels, and set the pod affinity/anti-affinity on all the pods. If both levels of pod affinity are set, then the leader pod will contain two pod affinities, while the workers will have a node selector, and a single pod affinity set. 
+
+### TPU Environment Variable Injection
+Because the value of the TPU environmental variables will vary between subgroups, the way they are injected will be extended. `TPU_WOKRER_ID` will be the value of `subgroup-worker-index`, while the value of `TPU_HOSTNAMES` will only be a list of the pods in the same subgroup.
```

Comment:
```suggestion
Because the value of the TPU environmental variables will vary between subgroups, the way they are injected will be extended. `TPU_WORKER_ID` will be the value of `subgroup-worker-index`, while the value of `TPU_HOSTNAMES` will only be a list of the pods in the same subgroup.
```

Reply:
fixed